flink-optimizer/src/main/java/org/apache/flink/optimizer/dag/OuterJoinNode.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.optimizer.dag;
+
+import org.apache.flink.api.common.functions.Partitioner;
+import org.apache.flink.api.common.operators.base.AbstractJoinOperatorBase.JoinHint;
+import org.apache.flink.api.common.operators.base.OuterJoinOperatorBase;
+import org.apache.flink.api.common.operators.base.OuterJoinOperatorBase.OuterJoinType;
+import org.apache.flink.optimizer.CompilerException;
+import org.apache.flink.optimizer.DataStatistics;
+import org.apache.flink.optimizer.operators.AbstractJoinDescriptor;
+import org.apache.flink.optimizer.operators.OperatorDescriptorDual;
+import org.apache.flink.optimizer.operators.SortMergeFullOuterJoinDescriptor;
+import org.apache.flink.optimizer.operators.SortMergeLeftOuterJoinDescriptor;
+import org.apache.flink.optimizer.operators.SortMergeRightOuterJoinDescriptor;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class OuterJoinNode extends TwoInputNode {
+
+	private List<OperatorDescriptorDual> dataProperties;
+
+	/**
+	 * Creates a new two input node for the optimizer plan, representing the given operator.
+	 *
+	 * @param operator The operator that the optimizer DAG node should represent.
+	 */
+	public OuterJoinNode(OuterJoinOperatorBase<?, ?, ?, ?> operator) {
+		super(operator);
+
+		this.dataProperties = getDataProperties();
+	}
+
+	private List<OperatorDescriptorDual> getDataProperties() {
+		OuterJoinOperatorBase<?, ?, ?, ?> operator = getOperator();
+
+		OuterJoinType type = operator.getOuterJoinType();
+
+		JoinHint joinHint = operator.getJoinHint();
+		joinHint = joinHint == null ? JoinHint.OPTIMIZER_CHOOSES : joinHint;
+
+		List<OperatorDescriptorDual> list = new ArrayList<>();
+		switch (joinHint) {
+			case OPTIMIZER_CHOOSES:
+				list.add(getSortMergeDescriptor(type, true));
+				list.add(getSortMergeDescriptor(type, false));
```

Comment:
For `OuterJoinType.FULL`, this will add the same descriptor a second time. This increases the number of enumerated plans and should be avoided.